### PR TITLE
add mpnn tutorial

### PR DESCRIPTION
**Pull request by akensert**
A graph neural network tutorial (specifically a message passing neural network for predicting graph properties)

### CODE REVIEW

**Comment by fchollet** [comment]
Thanks for the PR! Do you have a link to a Colab to preview the example?

File: examples/graph/mpnn-molecular-graphs.py [hunk 1]
```diff
@@ -0,0 +1,723 @@
...
+
+The functions below will help us generate graph features, namely, atom and bond features.
+Thanks to RDKit this can be done conventiently with only a few lines of code. Feel free
+to look up additional features that can be generated with
+[RDKit](https://www.rdkit.org/docs/Cookbook.html).
+"""
+
+
+def onehot_encode(x, allowable_set):
+    return list(map(lambda s: float(x == s), allowable_set))
```
Comment (fchollet) [inline examples/graph/mpnn-molecular-graphs.py:141]: This will iterate over the elements of the list every time you encode something. It's quite inefficient. Consider refactoring to use dicts/sets.

It seems you want to specify your allowed values as static lists. This is fine. But such lists will have to be turned into hashtables, and you only want to do this conversion once. So it sounds like you're going to want to use a class.

```python
class Index:
  def __init__(self, allowable_values):
    self.allowable_values = sorted(list(allowable_values))
    self.value_index = dict(zip(self.allowable_values, range(len(self.allowable_values))))

  def encode(self, value):
      if value not in self.value_index:
        raise ValueError(...)
      output = np.zeros((len(self.value_index),))
      output[self.value_index[value]] = 1.
      return output

atom_index = Index({"B", "Br", "C", "Ca", "Cl", "F", "I", "N", "Na", "O", "P", "S"})
output = atom_index.encode(atom.GetSymbol())
```

For practicality you may want to merge everything into a single big class that's aware of the atom semantics.

File: examples/graph/mpnn-molecular-graphs.py [hunk 2]
```diff
@@ -0,0 +1,723 @@
...
+y_test = df.iloc[test_index].p_np
+
+"""
+### Test out helper functions
+"""
+
+print(f"Name:\t{df.name[100]}\nSMILES:\t{df.smiles[100]}\nBBBP:\t{df.p_np[100]}")
+molecule = molecule_from_smiles(df.iloc[100].smiles)
+print("Molecule:")
+molecule
```
Comment (fchollet) [inline examples/graph/mpnn-molecular-graphs.py:304]: Move this to the print statement

File: examples/graph/mpnn-molecular-graphs.py [hunk 3]
```diff
@@ -0,0 +1,723 @@
...
+### Create tf.data.Dataset
+
+In this tutorial, the MPNN implementation will take as input (per iteration) a single graph.
+Therefore, given a batch of (sub-)graphs (molecules), we need to merge them into a
+single *global* graph. This global graph is a disconnected graph where each sub-graph is
+completely separated from the other sub-graphs.
+"""
+
+
+def prepare_batch(X_batch, y_batch):
```
Comment (fchollet) [inline examples/graph/mpnn-molecular-graphs.py:326]: Always use lowercase for variable names

File: examples/graph/mpnn-molecular-graphs.py [hunk 4]
```diff
@@ -0,0 +1,723 @@
...
+        atom_features_updated = tf.pad(atom_features, [(0, 0), (0, self.pad_length)])
+
+        # Perform a number of steps of message passing
+        for i in range(self.steps):
+
+            # Aggregate atom_features from neighbors
+            atom_features_aggregated = self.message_step(
+                [atom_features_updated, bond_features, pair_indices]
+            )
+            # # If no self-loops, zero-pad here.
```
Comment (fchollet) [inline examples/graph/mpnn-molecular-graphs.py:466]: Either remove the commented out code, or gate it via a `call()` argument.

File: examples/graph/mpnn-molecular-graphs.py [hunk 5]
```diff
@@ -0,0 +1,723 @@
...
+        memory_state = forget_gate * memory_state + update_gate * memory_state_candidate
+
+        # Update carry state
+        carry_state = output_gate * tf.nn.tanh(memory_state)
+
+        # Return (updated) memory state and carry state
+        return memory_state, carry_state
+
+
+class SetGather(keras.layers.Layer):
```
Comment (fchollet) [inline examples/graph/mpnn-molecular-graphs.py:546]: This is a very surprising architecture choice. The best tool for doing sequence-to-sequence over sets is to use a Transformer (without positional encoding). Transformer is the canonical set-processing building block.

File: examples/graph/mpnn-molecular-graphs.py [hunk 6]
```diff
@@ -0,0 +1,723 @@
...
+
+    atom_partition_indices = keras.layers.Input(
+        shape=(), dtype="int32", name="atom_partition_indices"
+    )
+
+    x = MessagePassing(message_units, message_steps)(
+        [atom_features, bond_features, pair_indices]
+    )
+
+    x = SetGather(batch_size, lstm_steps)([x, atom_partition_indices])
```
Comment (fchollet) [inline examples/graph/mpnn-molecular-graphs.py:652]: Try replacing this with a Transformer like https://keras.io/examples/nlp/neural_machine_translation_with_transformer/#building-the-model but without positional embedding

File: examples/graph/mpnn-molecular-graphs.py [hunk 7]
```diff
@@ -0,0 +1,723 @@
...
+    return model
+
+
+mpnn = MPNNModel(
+    atom_dim=X_train[0][0][0].shape[0], bond_dim=X_train[1][0][0].shape[0],
+)
+
+mpnn.compile(
+    loss=keras.losses.BinaryCrossentropy(),
+    optimizer=keras.optimizers.Adam(learning_rate=5e-4),
```
Comment (fchollet) [inline examples/graph/mpnn-molecular-graphs.py:671]: How did you obtain the learning rate?

File: examples/graph/mpnn-molecular-graphs.py [hunk 8]
```diff
@@ -0,0 +1,723 @@
...
+### Import packages
+"""
+
+import tensorflow as tf
+from tensorflow import keras
+import numpy as np
+import pandas as pd
+import matplotlib.pyplot as plt
+import warnings
+from rdkit import Chem, RDLogger
```
Comment (fchollet) [inline examples/graph/mpnn-molecular-graphs.py:93]: Only one import per line

File: examples/graph/mpnn-molecular-graphs.py [hunk 9]
```diff
@@ -0,0 +1,723 @@
...
+from rdkit.Chem.Draw import IPythonConsole, MolsToGridImage
+import logging
+
+tf.get_logger().setLevel(logging.ERROR)
+warnings.filterwarnings("ignore")
```
Comment (fchollet) [inline examples/graph/mpnn-molecular-graphs.py:98]: Are these 2 calls really necessary?

**Review by fchollet — COMMENTED** [review]
The code and explanations look great! This is a very interesting topic for an example.

File: examples/graph/mpnn-molecular-graphs.py (hunk 9 above)
Comment (akensert) [reply examples/graph/mpnn-molecular-graphs.py:98]: If commenting out `tf.get_logger().setLevel(logging.ERROR)`, I get the following warning when building or compiling the model:
```
WARNING:tensorflow:From /home/alex/Envs/keras_io/lib/python3.8/site-packages/tensorflow/python/ops/array_ops.py:5043: calling gather (from tensorflow.python.ops.array_ops) with validate_indices is deprecated and will be removed in a future version.
Instructions for updating:
The `validate_indices` argument has no effect. Indices are always validated on CPU and never validated on GPU.
```

And if commenting out `warnings.filterwarnings("ignore")`, I get the following type of warning(s) when fitting the model:
```
warnings.warn(
/home/alex/Envs/keras_io/lib/python3.8/site-packages/tensorflow/python/framework/indexed_slices.py:447: UserWarning: Converting sparse IndexedSlices(IndexedSlices(indices=Tensor("gradient_tape/model/message_passing/edge_network/Reshape_4:0", shape=(None,), dtype=int32), values=Tensor("gradient_tape/model/message_passing/edge_network/Reshape_3:0", shape=(None, 64), dtype=float32), dense_shape=Tensor("gradient_tape/model/message_passing/edge_network/Cast:0", shape=(2,), dtype=int32))) to a dense Tensor of unknown shape. This may consume a large amount of memory.
```

**Comment by akensert** [comment]
[Here's](https://colab.research.google.com/drive/1PDbhOb1tlrDtQ2jdnTdG3KAbKSJ-H08f?usp=sharing) a Google Colab for you to preview the example! Notice, I haven't yet updated it according to your feedback.

File: examples/graph/mpnn-molecular-graphs.py (hunk 1 above)
Comment (akensert) [reply examples/graph/mpnn-molecular-graphs.py:141]: Thank's a lot for the feedback, much appreciated. I'll look into it!

File: examples/graph/mpnn-molecular-graphs.py (hunk 2 above)
Comment (akensert) [reply examples/graph/mpnn-molecular-graphs.py:304]: The thing is, that the molecule will not be visualized if moved into the print statement (see Google Colab linked below)

File: examples/graph/mpnn-molecular-graphs.py (hunk 6 above)
Comment (akensert) [reply examples/graph/mpnn-molecular-graphs.py:652]: Thanks for helping out. I will study your tutorial and see what I can do. I think a transformer would be a great replacement. Notice however, that I tried to follow the original paper: Do you still think I should try to replace the set-to-set (*SetGather*) layer with a transformer?

File: examples/graph/mpnn-molecular-graphs.py (hunk 7 above)
Comment (akensert) [reply examples/graph/mpnn-molecular-graphs.py:671]: I first tried the default `1e-3` and it worked OK. However, lowering it to `5e-4` made the validation AUC (hence the learning curve) more stable.

File: examples/graph/mpnn-molecular-graphs.py (hunk 5 above)
Comment (akensert) [reply examples/graph/mpnn-molecular-graphs.py:546]: Yes indeed, I followed the [paper](https://arxiv.org/abs/1704.01212). Again, thanks a lot for the feedback, I'll look into it!

File: examples/graph/mpnn-molecular-graphs.py (hunk 6 above)
Comment (akensert) [reply examples/graph/mpnn-molecular-graphs.py:652]: I'm not entirely sure what the *inputs* should be to the decoder. I found [this paper](https://arxiv.org/pdf/1810.00825.pdf) which I think has an answer. They seem to use a learnable vector *S* (referred to as  *learnable set of k seed vectors* in the paper) as the *inputs*. What do you think? Do you have any other suggestions? What I want the transformer to output is a 2-D tensor (batch_size, embed_dim) which can be fed to the classification layers.

Another possible issue is that each example in the batch (which is merged into a big global graph before being fed to the MPNN) is of different size (i.e., the number of nodes differ between examples (molecules)). I'm thinking, should I obtain each example via `tf.dynamic_partition(global_graph, atom_partition_indices, ...)` and then pad each example to obtain a "rectangular" (3-D) tensor?

Comment (fchollet) [reply examples/graph/mpnn-molecular-graphs.py:652]: > Perhaps then the decoder could be excluded (so only encoder -> pooling -> dense), similar to this example (but without position encoding/embedding).

Yes, this is a fine solution. You can try it and see if you get improved results.

> Another possible issue is that each example in the batch (which is merged into a big global graph before being fed to the MPNN) is of different size

Didn't this issue already exist in the previous implementation? Your number of steps was hardcoded. Unless I missed something.

Using padding seems like a good idea. Remember to use masking so that the padded locations don't interfere with training.

**Comment by akensert** [comment]
Additional information on new commit (`update according to reviewer (now with black)`): (1) Replacing Set-to-set with a transformer encoder; (2) refactoring code for atom/bond feature generation. Also made minor changes according to your feedback.

Please check if the new implementations are OK. I think the masking is correct (tested it out offline), however, feel free to check it too. Thanks!

**Review by fchollet — APPROVED** [review]
Thanks for the update! Everything looks great. Glad to hear that using a transformer helped.

I pushed some copyedits, please pull them first. You can add the autogenerated files.

Random writing tip: avoid using the passive voice ("X was implemented") and prefer using the active voice, especially for your own actions ("we implemented X").

**Comment by akensert** [comment]
Thanks for the edits and the tip! I've committed the autogenerated files. Notice, I made some minor (additional) changes to the .py file.

**Comment by fchollet** [comment]
Thank you for the great contribution! 👍
